mailpoet/lib/Segments/DynamicSegments/Filters/WooCommercePurchaseDate.php

```diff
@@ -0,0 +1,73 @@
+<?php declare(strict_types = 1);
+
+namespace MailPoet\Segments\DynamicSegments\Filters;
+
+use MailPoet\Entities\DynamicSegmentFilterEntity;
+use MailPoet\Segments\DynamicSegments\Exceptions\InvalidFilterException;
+use MailPoetVendor\Doctrine\DBAL\Query\QueryBuilder;
+
+class WooCommercePurchaseDate implements Filter {
+  const ACTION = 'purchaseDate';
+
+  /** @var DateFilterHelper */
+  private $dateFilterHelper;
+
+  /** @var FilterHelper */
+  private $filterHelper;
+
+  /** @var WooFilterHelper */
+  private $wooFilterHelper;
+
+  public function __construct(
+    DateFilterHelper $dateFilterHelper,
+    FilterHelper $filterHelper,
+    WooFilterHelper $wooFilterHelper
+  ) {
+    $this->dateFilterHelper = $dateFilterHelper;
+    $this->filterHelper = $filterHelper;
+    $this->wooFilterHelper = $wooFilterHelper;
+  }
+
+  public function apply(QueryBuilder $queryBuilder, DynamicSegmentFilterEntity $filter): QueryBuilder {
+    $operator = $this->dateFilterHelper->getOperatorFromFilter($filter);
+    $dateValue = $this->dateFilterHelper->getDateValueFromFilter($filter);
+    $date = $this->dateFilterHelper->getDateStringForOperator($operator, $dateValue);
+    $subscribersTable = $this->filterHelper->getSubscribersTable();
+
+    if (in_array($operator, [DateFilterHelper::NOT_ON, DateFilterHelper::NOT_IN_THE_LAST])) {
+      $subQuery = $this->filterHelper->getNewSubscribersQueryBuilder();
+      $this->applyConditionsToQueryBuilder($operator, $date, $subQuery);
+      $queryBuilder->andWhere($queryBuilder->expr()->notIn("{$subscribersTable}.id", $subQuery->getSQL()));
+    } else {
+      $this->applyConditionsToQueryBuilder($operator, $date, $queryBuilder);
+    }
+
+    return $queryBuilder;
+  }
+
+  private function applyConditionsToQueryBuilder(string $operator, string $date, QueryBuilder $queryBuilder): QueryBuilder {
+    $orderStatsAlias = $this->wooFilterHelper->applyOrderStatusFilter($queryBuilder);
+    $quotedDate = $queryBuilder->expr()->literal($date);
```

Comment:
This is a tiny detail, but is there any reason for not using a placeholder (e.g., `?` or a named one) and `->setParameter(...)`? I guess Doctrine DBAL supports that as well. A similar question probably applies also [below](https://github.com/mailpoet/mailpoet/pull/4760/files#diff-c07f416fe577d2906c6a70e6d8b806ec4bbc39bd6711a0765c3ffe1eab7739afR80).

Reply:
Not a tiny detail at all! The reason I did it this way was because of the subquery and the fact that I'm calling `$subQuery->getSQL()`. With a prepared statement like you get with `setParameter`, the value doesn't come through when you call `getSQL()`, and you end up having to `setParameter` on the main query instead. I didn't like having to call `setParameter` like that, and I also didn't want to actually execute the subquery at this point, preventing any optimization that might occur at the DB level.

Using a placeholder seems like a much better way to go than my current solution, and I'll do that unless you think it would be better overall to use `setParameter`.

Honestly I haven't worked with doctrine very much, so any tips you have on best practices are very much appreciated!

Reply:
@JanJakes I took a stab at an approach that lets us use `setParameter` in 0c6614fc50e2ca6b8db4f9df34b2aadd7407ad1b. It still involves a string replacement, but only if we need to get the SQL string for the negated cases. It feels like there should be a more elegant solution and I plan on looking into it, but if you happen to have any advice I'm all ears!